File: backend/app/api/v1/app_version.py
"""Public app-version endpoint for forced/optional mobile updates.

Hit on cold start by the Flutter client. Public on purpose — the data
is non-sensitive (the Play Store URL is already public) and the
client needs the answer before it can decide whether to render the
login or home screen.
"""
from __future__ import annotations

from fastapi import APIRouter, Request

from app.core.config import settings
from app.schemas.app_version import AppVersionResponse

router = APIRouter()


def _pick_message(request: Request) -> str:
    """Pick a localised message based on the Accept-Language header.

    Falls back to UZ (the project's primary language) when the client
    doesn't advertise a supported locale.
    """
    header = (request.headers.get("accept-language") or "").lower()
    if header.startswith("en") and settings.app_update_message_en:
        return settings.app_update_message_en
    if header.startswith("ru") and settings.app_update_message_ru:
        return settings.app_update_message_ru
    return settings.app_update_message_uz


@router.get("/version", response_model=AppVersionResponse)
async def app_version(request: Request) -> AppVersionResponse:
    """Returns the current Play Store version + the minimum supported build.

    Mobile compares its own version+buildNumber to `min_supported_version`
    and to `force_update` to decide whether to show the blocking update
    dialog.
    """
    return AppVersionResponse(
        latest_version=settings.app_latest_version,
        min_supported_version=settings.app_min_supported_version,
        force_update=settings.app_force_update,
        update_url=settings.app_play_store_url,
        message=_pick_message(request),
    )

File: mobile/lib/models/app_version.dart
/// Backend-driven mobile app version config.
///
/// The server exposes these via `GET /api/v1/app/version` and the client
/// uses them on launch to decide whether to show a forced-update dialog.
/// Values are non-sensitive — the Play Store URL is public anyway.
class AppVersionConfig {
  const AppVersionConfig({
    required this.latestVersion,
    required this.minSupportedVersion,
    required this.forceUpdate,
    required this.updateUrl,
    this.message = '',
  });

  factory AppVersionConfig.fromJson(Map<String, dynamic> json) =>
      AppVersionConfig(
        latestVersion: (json['latest_version'] as String?) ?? '0.0.0',
        minSupportedVersion:
            (json['min_supported_version'] as String?) ?? '0.0.0',
        forceUpdate: json['force_update'] as bool? ?? false,
        updateUrl: (json['update_url'] as String?) ?? '',
        message: (json['message'] as String?) ?? '',
      );

  /// Newest build that's live on the Play Store (e.g. "1.2.0").
  final String latestVersion;

  /// Oldest build the server is willing to serve. Anything below this
  /// gets a non-dismissible update dialog.
  final String minSupportedVersion;

  /// Global kill switch — when true, every active build is told to
  /// update regardless of [minSupportedVersion].
  final bool forceUpdate;

  /// Play Store URL the dialog opens when the user taps "Update".
  final String updateUrl;

  /// Optional server-provided, already-localised message to show
  /// inside the dialog.
  final String message;

  /// Returns true when [current] (e.g. "1.0.0+5") is below the minimum
  /// supported version OR the server is forcing everyone to update.
  ///
  /// Trailing `+buildNumber` is stripped before comparison so the
  /// `package_info_plus` format ("1.0.0+5") compares against a pure
  /// semver ("1.0.0").
  bool requiresUpdate(String current) {
    if (forceUpdate) return true;
    final bare = current.split('+').first;
    return _compareVersions(bare, minSupportedVersion) < 0;
  }

  /// True when [current] is on a supported build but a newer one is
  /// available. This is a *soft* prompt — the user can dismiss it.
  /// Forced updates ([forceUpdate] or below [minSupportedVersion]) are
  /// never considered optional.
  bool hasOptionalUpdate(String current) {
    if (requiresUpdate(current)) return false;
    final bare = current.split('+').first;
    return _compareVersions(bare, latestVersion) < 0;
  }

  /// Compares two dot-separated version strings ("1.2.3" < "1.10.0").
  /// Returns negative if [a] < [b], zero if equal, positive if [a] > [b].
  /// Missing segments default to 0 so "1.2" == "1.2.0".
  static int _compareVersions(String a, String b) {
    final pa = a.split('.').map(int.tryParse).toList();
    final pb = b.split('.').map(int.tryParse).toList();
    final len = pa.length > pb.length ? pa.length : pb.length;
    for (var i = 0; i < len; i++) {
      final na = i < pa.length ? (pa[i] ?? 0) : 0;
      final nb = i < pb.length ? (pb[i] ?? 0) : 0;
      if (na != nb) return na - nb;
    }
    return 0;
  }
}
